plugins/Toolbox/src/CloudSync/SyncOrchestrator.py

```diff
@@ -0,0 +1,99 @@
+
+
+## Orchestrates the synchronizing of packages from the user account to the installed packages
+# Example flow:
+# - CloudPackageChecker compares a list of packages the user `subscribed` to in their account
+#   If there are `discrepancies` between the account and locally installed packages, they are emitted
+# - DiscrepanciesPresenter shows a list of packages to be added or removed to the user. It emits the `packageMutations`
+#   the user selected to be performed
+# - The SyncOrchestrator uses PackageManager to remove local packages the users wants to see removed
+# - The DownloadPresenter shows a download progress dialog. It emits A tuple of succeeded and failed downloads
+# - The LicensePresenter extracts licenses from the downloaded packages and presents a license for each package to
+#   be installed. It emits the `licenseAnswers` {'packageId' : bool} for accept or declines
+# - The CloudPackageManager removes the declined packages from the account
+# - The SyncOrchestrator uses PackageManager to install the downloaded packages.
+# - Bliss / profit / done
+class SyncOrchestrator(Extension):
+
+    def __init__(self, app: CuraApplication):
+        super().__init__()
+        # Differentiate This PluginObject from the Toolbox. self.getId() includes _name.
+        # getPluginId() will return the same value for The toolbox extension and this one
+        self._name = "SyncOrchestrator"
+
+        self._package_manager = app.getPackageManager()
+        self._cloud_package_manager = CloudPackageManager(app)
+
+        self._checker = CloudPackageChecker(app)  # type: CloudPackageChecker
+        self._checker.discrepancies.connect(self._onDiscrepancies)
+
+        self._discrepancies_presenter = DiscrepanciesPresenter(app)  # type: DiscrepanciesPresenter
+        self._discrepancies_presenter.packageMutations.connect(self._onPackageMutations)
+
+        self._download_Presenter = DownloadPresenter(app)  # type: DownloadPresenter
+
+        self._license_presenter = LicensePresenter(app)  # type: LicensePresenter
+        self._license_presenter.licenseAnswers.connect(self._onLicenseAnswers)
+
+        self._restart_presenter = RestartApplicationPresenter(app)
+
+    def _onDiscrepancies(self, model: SubscribedPackagesModel):
+        plugin_path = PluginRegistry.getInstance().getPluginPath(self.getPluginId())
+        self._discrepancies_presenter.present(plugin_path, model)
+
+    def _onPackageMutations(self, mutations: SubscribedPackagesModel):
+        self._download_Presenter = self._download_Presenter.resetCopy()
+        self._download_Presenter.done.connect(self._onDownloadFinished)
+        self._download_Presenter.download(mutations)
+
+    ## Called when a set of packages have finished downloading
+    # \param success_items: Dict[package_id, file_path]
+    # \param error_items: List[package_id]
+    def _onDownloadFinished(self, success_items: Dict[str, str], error_items: List[str]):
+        # todo handle error items
```

Comment:
If it is _not_ part of this ticket, then we're missing a ticket. We should either handle it in this one or create a new one.

As far as I can see now, all errors are written to the logs and/or appended to a list of erroneous packages. I'd like to see if things go wrong in the front-end. We should also have a think about what we should do with error'd packages in general... retry to download each time? Does it depend on what part of the download or installation went wrong?

Reply:
Added to the CCB